xwiki-platform-core/xwiki-platform-rendering/xwiki-platform-rendering-macros/xwiki-platform-rendering-macro-code-oldcore/src/main/java/org/xwiki/rendering/internal/macro/code/source/AbstractEntityCodeMacroSourceFactory.java

```diff
@@ -0,0 +1,140 @@
+        EntityReference entityReference =
+            this.resolver.resolve(reference.getReference(), getEntityType(), context.getCurrentMacroBlock());
+
+        XWikiContext xcontext = this.xcontextProvider.get();
+
+        if (xcontext == null) {
+            throw new MacroExecutionException("No XWiki context could be found in the current context");
+        }
+
+        // Resolve the document reference
+        DocumentReference documentReference = xcontext.getWiki().getDocumentReference(entityReference, xcontext);
+
+        // Current author must have view right on the document
+        try {
+            this.authorization.checkAccess(Right.VIEW, xcontext.getAuthorReference(), documentReference);
+        } catch (AccessDeniedException e) {
+            throw new MacroExecutionException(
+                "Current author is not allowed to access document [" + documentReference + "]");
+        }
+
+        // Current user must have view right on the document
+        if (this.authorization.hasAccess(Right.VIEW, xcontext.getUserReference(), documentReference)) {
+            throw new MacroExecutionException(
+                "Current user is not allowed to access document [" + documentReference + "]");
+        }
+
+        // Get the content
+        XWikiDocument document;
+        try {
+            document = xcontext.getWiki().getDocument(entityReference, xcontext);
+        } catch (XWikiException e) {
+            throw new MacroExecutionException("Failed to load document [" + documentReference + "]");
+        }
+
+        // Get the current translation
+        if (documentReference.getLocale() == null) {
+            try {
+                document = document.getTranslatedDocument(xcontext);
+            } catch (XWikiException e) {
+                throw new MacroExecutionException(
+                    "Failed to load document translation for reference [" + documentReference + "]");
+            }
+        }
+
+        return getContent(document, entityReference, reference, xcontext);
+    }
+
+    protected abstract EntityType getEntityType();
+
+    private CodeMacroSource getContent(XWikiDocument document, EntityReference entityReference,
+        CodeMacroSourceReference reference, XWikiContext xcontext) throws MacroExecutionException
+    {
+        if (!this.componentManager.hasComponent(EntityCodeMacroSourceLoader.class, entityReference.getType().name())) {
+            throw new MacroExecutionException(
+                "Unsupported entity type [" + entityReference.getType() + "] for reference [" + entityReference + "]");
+        }
+
+        EntityCodeMacroSourceLoader loader;
+        try {
+            loader =
+                this.componentManager.getInstance(EntityCodeMacroSourceLoader.class, entityReference.getType().name());
+        } catch (ComponentLookupException e) {
+            throw new MacroExecutionException("Unexpected error when initializing the content loader for entity type ["
```

Comment:
You don't log the original exception? Could be still useful no?

Reply:
Yes, just a miss.